File: c++ version/201-300/287-find-the-duplicate-number.cpp
int findDuplicate(vector<int>& nums) {
    int left = 1, right = nums.size();
    int mid = 0;
    int count = 0;
    while (left < right) {
        mid = left + (right-left)/2;
        count = 0;
        for (const auto &num : nums) {
            if (num <= mid) ++count;
        }
        if (count > mid) {
            right = mid;
        }
        else {
            left = mid + 1;
        }
    }
    return left;
}

File: python version/287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums)
        left, right = 1, n
        while left < right:
            mid = left + (right-left)//2
            count = 0
            for num in nums:
                if num <= mid: count += 1
            if count > mid:
                right = mid
            else:
                left = mid + 1
        return left